rasa/utils/tensorflow/transformer.py

```diff
@@ -0,0 +1,478 @@
+            if not self.relative_length:
+                raise ValueError(
+                    f"Max relative position {self.relative_length} "
+                    f"should be > 0 when using relative attention."
+                )
+
+            if self.unidirectional:
+                relative_length = self.relative_length
+            else:
+                relative_length = 2 * self.relative_length - 1
+
+            if self.heads_share_relative_embedding:
+                relative_embedding_shape = (relative_length, self._depth)
+            else:
+                relative_embedding_shape = (
+                    self.num_heads,
+                    relative_length,
+                    self._depth,
+                )
+        else:
+            relative_embedding_shape = None
+
+        if self.use_key_relative_position:
+            self.key_relative_embeddings = self.add_weight(
+                shape=relative_embedding_shape, name="key_relative_embeddings",
+            )
+        else:
+            self.key_relative_embeddings = None
+
+        if self.use_value_relative_position:
+            self.value_relative_embeddings = self.add_weight(
+                shape=relative_embedding_shape, name="value_relative_embeddings",
+            )
+        else:
+            self.value_relative_embeddings = None
+
+    def _pad_relative_embeddings(self, x: tf.Tensor, length: tf.Tensor) -> tf.Tensor:
+        # pad the left side to length
+        pad_left = x[:, :, :, :1, :]
+        pad_left = tf.tile(pad_left, (1, 1, 1, length - self.relative_length, 1))
+
+        # pad the right side to length
+        if self.unidirectional:
+            m_right = 1  # current time
+            pad_right = tf.zeros_like(x[:, :, :, -1:, :])
+        else:
+            m_right = self.relative_length
+            pad_right = x[:, :, :, -1:, :]
+        pad_right = tf.tile(pad_right, (1, 1, 1, length - m_right, 1))
+
+        return tf.concat([pad_left, x, pad_right], axis=-2)
+
+    def _slice_relative_embeddings(self, x: tf.Tensor, length: tf.Tensor) -> tf.Tensor:
+        if self.unidirectional:
+            # pad the right side to length
+            pad_right = tf.zeros_like(x[:, :, :, -1:, :])
+            pad_right = tf.tile(pad_right, (1, 1, 1, length - 1, 1))
+            x = tf.concat([x, pad_right], axis=-2)
+
+        dl = self.relative_length - length
+        m = tf.shape(x)[-2]
+        return x[:, :, :, dl : m - dl, :]
+
+    def _relative_to_absolute_position(self, x: tf.Tensor) -> tf.Tensor:
+        """Universal method to convert tensor from relative to absolute indexing.
+
+        x.shape =
+        (batch, num_heads, length, relative_length, depth)
+        or (batch, num_heads, length, relative_length)
+        "Slides" relative embeddings by 45 degree """
+
+        x_dim = len(x.shape)
+
+        if x_dim < 4 or x_dim > 5:
+            raise ValueError("Relative tensor has a wrong shape.")
```

Comment:
Should we mention (1) the shape and (2) state that it should be either of dimensionality 4 or 5?

Reply:
it's more of an internal method, I added error not make mistake myself